### DISCUSSION TITLE
Automated generation of blank nodes without templates or references

### DISCUSSION THREAD

**Discussion by lucascool12** [morph-kgc/morph-kgc#107] [Ideas]
####  What Happens?
A triples map that has a subject map with a blanknode as term type causes an exception if there is no rr:template in the subject map.

#### To Reproduce
Steps to reproduce the behavior. If possible include sample data, mappings and configuration used.
A triples map with a subject map as blanknode without rr:template in subject map.
[blankNodeExample.csv](https://github.com/oeg-upm/morph-kgc/files/9862608/blankNodeExample.csv)
Mapping:
```
@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
@prefix rr: <http://www.w3.org/ns/r2rml#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/>.

_:b a rr:TriplesMap;
    rml:logicalSource[
        rml:referenceFormulation <http://semweb.mmlab.be/ns/ql#CSV>;
        rml:source "test.csv"
    ];
    rr:subjectMap [
        # rr:template "{id}";
        rr:termType rr:BlankNode
    ];
    rr:predicateObjectMap [
        rr:predicate foaf:firstName;
        rr:objectMap [
            rml:reference "given_name"
        ]
    ].
```
Config:
[DataSourceCSV]
mappings=blankNodeExample.ttl

#### Environment (please complete the following information):
 - OS: Ubuntu 20.04.5 LTS on Windows 10 x86_64
 - Python version: 3.10.8
 - Morph-KGC version: 2.2.0

**Comment by arenas-guerrero-julian** [comment]
Hi @lucascool12 

As far as I know, the behaviour of morph-kgc is the standard of the [R2RML Recommendation](https://www.w3.org/TR/r2rml/). A term map must be provided a constant, column/reference or a template.

However, I think the [rmlmapper](https://github.com/RMLio/rmlmapper-java) does support this mapping syntax as an extended functionality. Maybe @DylanVanAssche can confirm this?

Why do you need this functionality? What is the limitation of standard [R2]RML for your use case?

Julián

**Comment by lucascool12** [comment]
As you said rmlmapper does support this, that plus the incredible stack trace the example above produces let me down the wrong path.
This functionality would be convenient, in the same style as [issue 78](https://github.com/oeg-upm/morph-kgc/issues/78).
As I now know that this is not part of standard [R2]RML, I won't be using it.
So i'll close this issue.

Thank you for your time.

**Comment by dachafra** [comment]
In any case, it's good to mention that the W3C community group on KGC agreed that in the next version of RML this feature will be allowed ([see here](https://github.com/kg-construct/rml-core/issues/52)), but this will require extra effort to implement it as references would be maintained in memory if there is any join.
